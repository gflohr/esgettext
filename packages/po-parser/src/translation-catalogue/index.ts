export * from './translation-catalogue';
export * from './po-entry';
