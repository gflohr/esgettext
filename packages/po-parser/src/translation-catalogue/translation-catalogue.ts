import { PoEntry } from './po-entry';

// FIXME! We don't a cache here!
interface Cache {
	// msgid
	[key: string]: {
		// msgctxt
		[key: string]: PoEntry;
	};
}

export type RenderOptions = {
	width?: number;
	sortOutput?: boolean;
	sortByFile?: boolean;
	omitHeader?: boolean;
};

export type POHeader = {
	ProjectIdVersion: string;
	ReportMsgidBugsTo: string;
	POTCreationDate: string;
	PORevisionDate: string;
	LastTraslator: string;
	LanguageTeam: string;
	Language: string;
	MIMEVersion: string;
	ContentType: string;
	ContentTransferEncoding: string;
	PluralForms: string;
	[field: string]: string;
};

/**
 * An entire message catalog that is subsequently filled.
 */
export class TranslationCatalogue {
	private readonly cache: Cache = {};
	entries: Array<PoEntry>;

	constructor() {
		this.entries = new Array<PoEntry>();
	}

	/**
	 * Render the entire catalog as a PO file.
	 *
	 * @param renderOptions - options for rendering
	 *
	 * The following options are supported:
	 *
	 * width: wrap lines to that page width (default: 76)
	 *
	 * @returns the catalog rendered as a po(t) file
	 */
	renderPo(options: RenderOptions = {}): string {
		const width =
			options && typeof options.width !== 'undefined' ? options.width : 76;

		const isNotHeader = (entry: PoEntry): boolean => {
			return !(
				entry.properties.msgid === '' &&
				typeof entry.properties.msgctxt === 'undefined'
			);
		};
		const isHeader = (entry: PoEntry): boolean => {
			return !isNotHeader(entry);
		};

		const header = options.omitHeader
			? undefined
			: this.entries
					.filter(isHeader)
					.map(entry => entry.toString(width))
					.join('\n');

		let body = '';
		if (options.sortOutput) {
			body = this.entries
				.filter(isNotHeader)
				.sort((a, b) => {
					const cmp = a.properties.msgid.localeCompare(b.properties.msgid);
					if (cmp) {
						return cmp;
					}
					const actx = a.properties.msgctxt;
					const bctx = b.properties.msgctxt;

					// It is impossible that both msgctxts are undefined
					// because that would mean that the msgids are identical
					// which is prevented.
					if (typeof bctx === 'undefined') {
						return +1;
					} else if (typeof actx === 'undefined') {
						return -1;
					}

					return actx.localeCompare(bctx);
				})
				.map(entry => entry.toString(width))
				.join('\n');
		} else if (options.sortByFile) {
			const copy = this.copy();

			const splitRef = (ref: string): { filename: string; lineno: number } => {
				const parts = ref.split(':');
				const lineno = Number.parseInt(parts.pop() as string, 10);
				const filename = parts.join(':');
				return { filename, lineno };
			};

			// First order all references
			copy.entries.forEach(entry => {
				if (entry.properties.references) {
					entry.properties.references = entry.properties.references
						.map(splitRef)
						.sort((a, b) => {
							return (
								a.filename.localeCompare(b.filename) ||
								Math.sign(a.lineno - b.lineno)
							);
						})
						.map(ref => `${ref.filename}:${ref.lineno}`);
				}
			});

			const compareSplitRefs = (
				a: {
					refs: Array<{ filename: string; lineno: number }>;
					entry: PoEntry;
				},
				b: {
					refs: Array<{ filename: string; lineno: number }>;
					entry: PoEntry;
				},
			): number => {
				const arefs = a.refs;
				const brefs = b.refs;
				const min = Math.min(arefs.length, brefs.length);
				for (let i = 0; i < min; ++i) {
					const cmp =
						arefs[i].filename.localeCompare(brefs[i].filename) ||
						Math.sign(arefs[i].lineno - brefs[i].lineno);
					if (cmp) {
						return cmp;
					}
				}
				return Math.sign(arefs.length - brefs.length);
			};

			const splitRefs = (
				refs: Array<string>,
			): Array<{ filename: string; lineno: number }> => {
				return refs ? refs.map(splitRef) : [];
			};

			body = copy.entries
				.filter(isNotHeader)
				.map(entry => {
					return {
						refs: splitRefs(entry.properties.references as string[]),
						entry,
					};
				})
				.sort(compareSplitRefs)
				.map(split => split.entry)
				.map(entry => entry.toString(width))
				.join('\n');
		} else {
			body = this.entries
				.filter(isNotHeader)
				.map(entry => entry.toString(width))
				.join('\n');
		}

		if (body && header) {
			return `${header}\n${body}`;
		} else {
			return header || body || '';
		}
	}

	/**
	 * Add an entry to the catalog. If an entry with the same msgid
	 * and msgctxt already exists, the entry is merged into
	 * the existing one instead.
	 *
	 * @param entry - the `PoEntry` to add
	 */
	addEntry(entry: PoEntry): void {
		const msgid = entry.properties.msgid;
		const msgctxt = entry.properties.msgctxt as string;

		if (!Object.prototype.hasOwnProperty.call(this.cache, msgid)) {
			this.cache[msgid] = {};
		}
		if (!Object.prototype.hasOwnProperty.call(this.cache[msgid], msgctxt)) {
			this.cache[msgid][msgctxt] = entry;
			this.entries.push(entry);
		} else {
			this.cache[msgid][msgctxt].merge(entry);
		}
	}

	/**
	 * Remove an entry from the catalog.
	 *
	 * @param msgid - the message id
	 * @param msgctxt - the message context
	 */
	deleteEntry(msgid: string, msgctxt?: string) {
		if (!this.cache[msgid]) {
			return;
		}

		const entry = this.cache[msgid][msgctxt as string];
		if (!entry) {
			return;
		}

		delete this.cache[msgid][msgctxt as string];
		this.entries = this.entries.filter(other => other !== entry);
	}
	/**
	 * Add an entry to the catalog. If an entry with the same msgid
	 * and msgctxt already exists, the entry is merged into
	 * the existing one instead.
	 *
	 * @param entry - the `PoEntry` to add
	 */
	getEntry(msgid: string, msgctxt?: string): PoEntry | undefined {
		return this.entries.find(entry => {
			return (
				entry.properties.msgid === msgid && entry.properties.msgctxt === msgctxt
			);
		});
	}

	/**
	 * Copy a catalogue.
	 *
	 * @returns a deep copy of the catalogue.
	 */
	copy(): TranslationCatalogue {
		const other = new TranslationCatalogue();

		this.entries.forEach(entry => other.addEntry(entry));

		return other;
	}

	/**
	 * Removes all translations from a catalog, so that it can be used as
	 * a .pot file.
	 */
	makePOT(): void {
		this.entries.forEach(entry => {
			if (
				!(
					entry.properties.msgid === '' &&
					typeof entry.properties.msgctxt === 'undefined'
				)
			) {
				delete entry.properties.msgstr;
			}
		});
	}
}
