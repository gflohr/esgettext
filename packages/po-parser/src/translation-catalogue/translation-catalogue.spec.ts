import { TranslationCatalogue } from './translation-catalogue';
import { PoEntry } from './po-entry';

describe('translation catalogue', () => {
	describe('initialization', () => {
		it('should be initialized with zero configuration', () => {
			const catalogue = new TranslationCatalogue();
			expect(catalogue.renderPo()).toBe('');
		});
	});

	describe('conflicting flags', () => {
		/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
		let warner: jest.SpyInstance<void, [any?, ...any[]]>;

		beforeEach(() => {
			warner = jest.spyOn(global.console, 'warn').mockImplementation(() => {
				/* ignore */
			});
		});

		afterEach(() => jest.restoreAllMocks());

		it('should warn about conflicting flags', () => {
			const catalogue = new TranslationCatalogue();

			catalogue.addEntry(
				new PoEntry({
					msgid: 'hello',
					flags: ['no-mercy', 'perl-brace-format'],
					references: ['first.js:10'],
				}),
			);

			catalogue.addEntry(
				new PoEntry({
					msgid: 'hello',
					flags: ['no-mercy'],
					references: ['second.js:10'],
				}),
			);

			catalogue.addEntry(
				new PoEntry({
					msgid: 'hello',
					flags: ['mercy'],
					references: ['third.js:10'],
				}),
			);
			expect(warner).toHaveBeenCalledTimes(4);

			catalogue.addEntry(
				new PoEntry({
					msgid: 'hello',
					flags: ['no-perl-brace-format'],
					references: ['third.js:10'],
				}),
			);
			expect(warner).toHaveBeenCalledTimes(9);
		});
	});

	describe('continuous filling', () => {
		const catalogue = new TranslationCatalogue();

		it('should add an entry', () => {
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Mike',
					references: ['src/zulu.ts:2304'],
				}),
			);
			expect(catalogue.renderPo()).toMatchSnapshot();
		});

		it('should add a second entry', () => {
			catalogue.addEntry(
				new PoEntry({
					msgid: 'November',
					references: ['src/yankee.ts:42'],
				}),
			);
			expect(catalogue.renderPo()).toMatchSnapshot();
		});

		it('should merge the third entry', () => {
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Mike',
					references: ['src/xray.ts:1303'],
				}),
			);
			expect(catalogue.renderPo()).toMatchSnapshot();
		});

		// This can later be used to test sorting by file.
		it('should support a string at many locations', () => {
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Lima',
					references: ['whiskey.ts:2'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Oscar',
					references: ['whiskey.ts:2'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Lima',
					references: ['victor.ts:3'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Oscar',
					references: ['victor.ts:3'],
				}),
			);
			// When sorting by file, the Oscar entry should appear before
			// the Lima entry. Actually that case could not happen but ...
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Lima',
					references: ['uniform.ts:4'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Oscar',
					references: ['uniform.ts:3'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Lima',
					references: ['tango.ts:25'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Oscar',
					references: ['tango.ts:25'],
				}),
			);
			expect(catalogue.renderPo()).toMatchSnapshot();
		});
		it('should also merge flags and comments', () => {
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Mike',
					translatorComments: ['This is difficult.'],
					automatic: ['TRANSLATORS: Buy me a milk shake!'],
					flags: ['no-c-format'],
					references: ['sierra.ts:27'],
				}),
			);
			expect(catalogue.renderPo()).toMatchSnapshot();

			catalogue.addEntry(
				new PoEntry({
					msgid: 'Mike',
					translatorComments: ['Not really.'],
					automatic: ['TRANSLATORS: Drop me a line!'],
					flags: ['perl-brace-format'],
					references: ['romeo.ts:85'],
				}),
			);
			expect(catalogue.renderPo()).toMatchSnapshot();

			// We need entries with a message context.
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Mike',
					msgctxt: 'phonetic alphabet',
					references: ['juliet.ts:85'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Mike',
					msgctxt: 'audio',
					references: ['golf.ts:85'],
				}),
			);

			// And entries with one more ref than another.
			catalogue.addEntry(
				new PoEntry({
					msgid: 'Foxtrot',
					references: ['src/yankee.ts:42', 'src/zulu.ts:2304'],
				}),
			);
			expect(catalogue).toMatchSnapshot();
		});

		it('should sort output on demand', () => {
			expect(catalogue.renderPo({ sortOutput: true })).toMatchSnapshot();
		});

		it('should sort by file on demand', () => {
			// The snapshot differs from the GNU xgettext output but I think
			// that our version is correct, and the "original" is wrong.
			// Anyway, who needs that feature???
			expect(catalogue.renderPo({ sortByFile: true })).toMatchSnapshot();
		});
	});

	describe('deleting', () => {
		const catalogue = new TranslationCatalogue();

		catalogue.addEntry(
			new PoEntry({
				msgid: 'Hello, world!',
			}),
		);

		expect(catalogue.renderPo()).toMatchSnapshot();

		catalogue.deleteEntry('Hello, world!');
		catalogue.deleteEntry('Goodbye, world!');
		catalogue.deleteEntry('Hello, world!', 'La-la-land');

		expect(catalogue.renderPo()).toMatchSnapshot();

		catalogue.deleteEntry('');
		expect(catalogue.renderPo()).toEqual('');
	});

	describe('sorting', () => {
		it('should sort by msgctxt', () => {
			const catalogue = new TranslationCatalogue();

			catalogue.addEntry(
				new PoEntry({
					msgid: 'sort',
					msgctxt: 'delta',
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'sort',
					msgctxt: 'charlie',
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'sort',
					msgctxt: 'bravo',
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'sort',
					msgctxt: 'alpha',
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'sort',
				}),
			);

			expect(catalogue.renderPo({ sortOutput: true })).toMatchSnapshot();
		});

		it('should sort by line number', () => {
			const catalogue = new TranslationCatalogue();

			catalogue.addEntry(
				new PoEntry({
					msgid: 'alpha',
					references: ['source.ts:42'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'bravo',
					references: ['source.ts:3'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'charlie',
					references: ['source.ts:31'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'delta',
					references: ['source.ts:1'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'echo',
					references: ['source.ts:11', 'source.ts:13'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'foxtrot',
					references: ['source.ts:11', 'source.ts:12'],
				}),
			);
			catalogue.addEntry(
				new PoEntry({
					msgid: 'golf',
				}),
			);

			expect(catalogue.renderPo({ sortByFile: true })).toMatchSnapshot();
		});
	});
});
