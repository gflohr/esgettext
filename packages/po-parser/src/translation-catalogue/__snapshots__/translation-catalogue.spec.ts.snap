// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"msgid "Hello, world!"
msgstr ""
"
`;

exports[` 2`] = `""`;

exports[`translation catalogue continuous filling should add a second entry 1`] = `
"#: src/zulu.ts:2304
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should add an entry 1`] = `
"#: src/zulu.ts:2304
msgid "Mike"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should also merge flags and comments 1`] = `
"# This is difficult.
#. TRANSLATORS: Buy me a milk shake!
#: src/zulu.ts:2304 src/xray.ts:1303 sierra.ts:27
#, no-c-format
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid "Lima"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid "Oscar"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should also merge flags and comments 2`] = `
"# This is difficult.
# Not really.
#. TRANSLATORS: Buy me a milk shake!
#. TRANSLATORS: Drop me a line!
#: src/zulu.ts:2304 src/xray.ts:1303 sierra.ts:27 romeo.ts:85
#, no-c-format, perl-brace-format
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid "Lima"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid "Oscar"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should also merge flags and comments 3`] = `
TranslationCatalogue {
  "cache": {
    "Foxtrot": {
      "undefined": PoEntry {
        "properties": {
          "msgid": "Foxtrot",
          "references": [
            "src/yankee.ts:42",
            "src/zulu.ts:2304",
          ],
        },
      },
    },
    "Lima": {
      "undefined": PoEntry {
        "properties": {
          "msgid": "Lima",
          "references": [
            "whiskey.ts:2",
            "victor.ts:3",
            "uniform.ts:4",
            "tango.ts:25",
          ],
        },
      },
    },
    "Mike": {
      "audio": PoEntry {
        "properties": {
          "msgctxt": "audio",
          "msgid": "Mike",
          "references": [
            "golf.ts:85",
          ],
        },
      },
      "phonetic alphabet": PoEntry {
        "properties": {
          "msgctxt": "phonetic alphabet",
          "msgid": "Mike",
          "references": [
            "juliet.ts:85",
          ],
        },
      },
      "undefined": PoEntry {
        "properties": {
          "automatic": [
            "TRANSLATORS: Buy me a milk shake!",
            "TRANSLATORS: Drop me a line!",
          ],
          "flags": [
            "no-c-format",
            "perl-brace-format",
          ],
          "msgid": "Mike",
          "references": [
            "src/zulu.ts:2304",
            "src/xray.ts:1303",
            "sierra.ts:27",
            "romeo.ts:85",
          ],
          "translatorComments": [
            "This is difficult.",
            "Not really.",
          ],
        },
      },
    },
    "November": {
      "undefined": PoEntry {
        "properties": {
          "msgid": "November",
          "references": [
            "src/yankee.ts:42",
          ],
        },
      },
    },
    "Oscar": {
      "undefined": PoEntry {
        "properties": {
          "msgid": "Oscar",
          "references": [
            "whiskey.ts:2",
            "victor.ts:3",
            "uniform.ts:3",
            "tango.ts:25",
          ],
        },
      },
    },
  },
  "entries": [
    PoEntry {
      "properties": {
        "automatic": [
          "TRANSLATORS: Buy me a milk shake!",
          "TRANSLATORS: Drop me a line!",
        ],
        "flags": [
          "no-c-format",
          "perl-brace-format",
        ],
        "msgid": "Mike",
        "references": [
          "src/zulu.ts:2304",
          "src/xray.ts:1303",
          "sierra.ts:27",
          "romeo.ts:85",
        ],
        "translatorComments": [
          "This is difficult.",
          "Not really.",
        ],
      },
    },
    PoEntry {
      "properties": {
        "msgid": "November",
        "references": [
          "src/yankee.ts:42",
        ],
      },
    },
    PoEntry {
      "properties": {
        "msgid": "Lima",
        "references": [
          "whiskey.ts:2",
          "victor.ts:3",
          "uniform.ts:4",
          "tango.ts:25",
        ],
      },
    },
    PoEntry {
      "properties": {
        "msgid": "Oscar",
        "references": [
          "whiskey.ts:2",
          "victor.ts:3",
          "uniform.ts:3",
          "tango.ts:25",
        ],
      },
    },
    PoEntry {
      "properties": {
        "msgctxt": "phonetic alphabet",
        "msgid": "Mike",
        "references": [
          "juliet.ts:85",
        ],
      },
    },
    PoEntry {
      "properties": {
        "msgctxt": "audio",
        "msgid": "Mike",
        "references": [
          "golf.ts:85",
        ],
      },
    },
    PoEntry {
      "properties": {
        "msgid": "Foxtrot",
        "references": [
          "src/yankee.ts:42",
          "src/zulu.ts:2304",
        ],
      },
    },
  ],
}
`;

exports[`translation catalogue continuous filling should merge the third entry 1`] = `
"#: src/zulu.ts:2304 src/xray.ts:1303
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should sort by file on demand 1`] = `
"#: golf.ts:85
msgctxt "audio"
msgid "Mike"
msgstr ""

#: juliet.ts:85
msgctxt "phonetic alphabet"
msgid "Mike"
msgstr ""

# This is difficult.
# Not really.
#. TRANSLATORS: Buy me a milk shake!
#. TRANSLATORS: Drop me a line!
#: romeo.ts:85 sierra.ts:27 src/xray.ts:1303 src/zulu.ts:2304
#, no-c-format, perl-brace-format
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""

#: src/yankee.ts:42 src/zulu.ts:2304
msgid "Foxtrot"
msgstr ""

#: tango.ts:25 uniform.ts:3 victor.ts:3 whiskey.ts:2
msgid "Oscar"
msgstr ""

#: tango.ts:25 uniform.ts:4 victor.ts:3 whiskey.ts:2
msgid "Lima"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should sort output on demand 1`] = `
"#: src/yankee.ts:42 src/zulu.ts:2304
msgid "Foxtrot"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid "Lima"
msgstr ""

# This is difficult.
# Not really.
#. TRANSLATORS: Buy me a milk shake!
#. TRANSLATORS: Drop me a line!
#: src/zulu.ts:2304 src/xray.ts:1303 sierra.ts:27 romeo.ts:85
#, no-c-format, perl-brace-format
msgid "Mike"
msgstr ""

#: golf.ts:85
msgctxt "audio"
msgid "Mike"
msgstr ""

#: juliet.ts:85
msgctxt "phonetic alphabet"
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid "Oscar"
msgstr ""
"
`;

exports[`translation catalogue continuous filling should support a string at many locations 1`] = `
"#: src/zulu.ts:2304 src/xray.ts:1303
msgid "Mike"
msgstr ""

#: src/yankee.ts:42
msgid "November"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid "Lima"
msgstr ""

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid "Oscar"
msgstr ""
"
`;

exports[`translation catalogue sorting should sort by line number 1`] = `
"msgid "golf"
msgstr ""

#: source.ts:1
msgid "delta"
msgstr ""

#: source.ts:3
msgid "bravo"
msgstr ""

#: source.ts:11 source.ts:12
msgid "foxtrot"
msgstr ""

#: source.ts:11 source.ts:13
msgid "echo"
msgstr ""

#: source.ts:31
msgid "charlie"
msgstr ""

#: source.ts:42
msgid "alpha"
msgstr ""
"
`;

exports[`translation catalogue sorting should sort by msgctxt 1`] = `
"msgid "sort"
msgstr ""

msgctxt "alpha"
msgid "sort"
msgstr ""

msgctxt "bravo"
msgid "sort"
msgstr ""

msgctxt "charlie"
msgid "sort"
msgstr ""

msgctxt "delta"
msgid "sort"
msgstr ""
"
`;
